app/src/main/java/br/org/cesar/madmovies/movies/domain/usecase/GetMovieDetail.kt

```diff
@@ -0,0 +1,12 @@
+package br.org.cesar.madmovies.movies.domain.usecase
+
+import android.util.Log
+import br.org.cesar.madmovies.movies.data.Repository
+import br.org.cesar.madmovies.movies.domain.model.Movie
+
+object GetMovieDetail {
```

Comment:
I'm finding this redundant. Is this a use case? If so, why not a class (instead of object) receiving the repository in constructor (instead of passing the repository to invoke function)?

Anyway, not a fan of "proxy" use cases (use cases that do nothing except calling the repository, usually a method with the same name), but I understand the appeal. I'd certainly receive the repository in class constructor, though.

Reply:
We have chosen this approach because the UseCase should be a singleton. It does not make sense to have multiple instances of the same use case. If we need to get the details of a movie from a cache repository or a remote repository, in both cases the logic is the same, call the method from said repository. It does not make sense to have to create a separated instance of said use case if we want to have it executed in different repositories.

That's the reason behind the choice

Reply:
"Cache" or "remote" are data sources for the same repository

Still, if you want to mock a repository for unit testing the use case, you need to pass it as a constructor parameter to the use case. That makes more sense than requiring passing in the repository at every call to the use case method.

So the best pattern here would not be a pure singleton, but a "parametric singleton" -- which would be a normal class where the "singletonness" is enforced by the DI, or by the factory (think of viewModels: they are "singletons" for any given ViewModelStore, and they can take parameters)

Unfortunately Kotlin does not have a language construct for "parametric singletons" so this must be handled by the programmer. This can be implemented with DI or with the Factory pattern.